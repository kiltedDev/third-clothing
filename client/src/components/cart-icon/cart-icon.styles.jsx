import styled from 'styled-components';

import Icon from '../../icons/icon.index';

export const CartIconContainer = styled.div`
  width: 45px;
  height: 45px;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;

  svg {
    height: 100%;
    width: 100%;
  }
`

export const ShoppingIcon = styled(Icon)`
  width: 24px;
  height: 24px;
`

export const ItemCount = styled.span`
  position: absolute;
  font-size: 10px;
  font-weight: bold;
  bottom: 12px;
`

CartIconContainer.displayName = 'CartIconContainer';
ShoppingIcon.displayName = 'ShoppingIcon';
ItemCount.displayName = 'ItemCount';
