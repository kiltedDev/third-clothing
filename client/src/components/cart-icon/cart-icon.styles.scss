.cart-icon {
  width: 45px;
  height: 45px;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;

  .shopping-icon {
    width: 24px;
    height: 24px;
  }

  svg {
    height: 100%;
    width: 100%;
  }
  
  .item-count {
    position: absolute;
    font-size: 10px;
    font-weight: bold;
    bottom: 12px;
  }
}
