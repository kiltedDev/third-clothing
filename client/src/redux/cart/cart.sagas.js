import { all, call, put, takeLatest, select } from 'redux-saga/effects';

import UserActionTypes from '../user/user.types';
import CartActionTypes from './cart.types';

import { clearCart, setCartFromFirebase } from '../cart/cart.actions';
import { getUserCartRef, newUserCart } from '../../firebase/firebase.utils';
import { selectCurrentUser } from '../user/user.selectors';
import { selectCartItems } from './cart.selectors';

export function* clearCartOnSignOut() {
  yield put( clearCart() );
}

export function* onSignOutSuccess() {
  yield takeLatest( UserActionTypes.SIGN_OUT_SUCCESS, clearCartOnSignOut )
}

export function* updateCartInFirebase() {
  const currentUser = yield select( selectCurrentUser );
  if (currentUser) {
    try {
      const cartRef = yield getUserCartRef( currentUser.id, currentUser.activeCart );
      const cartItems = yield select( selectCartItems );
      yield cartRef.update({ cartItems });
    } catch ( error ) {
      console.log( error.message );
    }
  }
}

export function* checkCartFromFirebase({ payload: user }) {
  const cartRef = yield getUserCartRef( user.id, user.activeCart );
  const cartSnapshot = yield cartRef.get();
  yield put( setCartFromFirebase( cartSnapshot.data().cartItems ) );
}

export function* createNewUserCart() {
  const currentUser = yield select( selectCurrentUser );
  if ( currentUser ) {
    try {
      console.log(currentUser.id);
      const newCartRef = yield newUserCart( currentUser.id );
      console.log(newCartRef);
    } catch ( error ) {
      console.log( error.message )
    }
  }
  yield put( setCartFromFirebase( [] ) );

}

export function* onUserSignIn() {
  yield takeLatest(UserActionTypes.SIGN_IN_SUCCESS, checkCartFromFirebase);
}

export function* onCartChange() {
  yield takeLatest(
    [
      CartActionTypes.ADD_ITEM,
      CartActionTypes.REMOVE_ITEM,
      CartActionTypes.CLEAR_ITEM_FROM_CART
    ],
    updateCartInFirebase
  );
}

export function* onCheckOut() {
  yield takeLatest(
    [
      CartActionTypes.CHECK_OUT_CART
    ],
    createNewUserCart
  );
}

export function* cartSagas() {
  yield( all([
    call( onSignOutSuccess ),
    call( onCartChange ),
    call( onUserSignIn ),
    call( onCheckOut )
  ]))
}
