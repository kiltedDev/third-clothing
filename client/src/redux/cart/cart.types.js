export const CartActionTypes = {
  TOGGLE_CART_HIDDEN: 'TOGGLE_CART_HIDDEN',
  HIDE_CART: 'HIDE_CART',
  ADD_ITEM: 'ADD_ITEM',
  REMOVE_ITEM: 'REMOVE_ITEM',
  CLEAR_ITEM_FROM_CART: 'CLEAR_ITEM_FROM_CART',
  CLEAR_CART: 'CLEAR_CART',
  UPDATE_CART_IN_FIREBASE: 'UPDATE_CART_IN_FIREBASE',
  SET_CART_FROM_FIREBASE: 'SET_CART_FROM_FIREBASE'
};

export default CartActionTypes;
