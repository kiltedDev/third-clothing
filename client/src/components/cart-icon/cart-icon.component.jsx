import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { toggleCartHidden } from '../../redux/cart/cart.actions';
import { selectCartItemsCount } from '../../redux/cart/cart.selectors';

import Icon from '../../icons/icon.index';

import {
  CartIconContainer,
  ShoppingIcon,
  ItemCount
} from './cart-icon.styles';

export const CartIcon = ({ itemCount, toggleCartHidden }) => (
  <CartIconContainer
    className='cart-icon'
    onClick={ toggleCartHidden }
    >
    <ShoppingIcon
      className='shopping-icon'
      title='shopping-bag'
    />
    <ItemCount
      className='item-count'
      children={ itemCount }
    />
  </CartIconContainer>
)

const mapStateToProps = createStructuredSelector({
  itemCount: selectCartItemsCount
})

const mapDispatchToProps = dispatch => ({
  toggleCartHidden: () => dispatch( toggleCartHidden() )
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CartIcon);
