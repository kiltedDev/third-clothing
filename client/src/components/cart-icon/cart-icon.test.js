import { shallow } from 'enzyme';
import React from 'react';
import { CartIcon } from './cart-icon.component';

import { toggleCartHidden } from '../../redux/cart/cart.actions'

describe('CartIcon', () => {
  let wrapper;
  let mockToggleCartHidden;

  beforeAll(() => {
    mockToggleCartHidden = jest.fn();
    const mockProps = {
      itemCount: 0,
      toggleCartHidden: mockToggleCartHidden
    }

    wrapper = shallow(<CartIcon { ...mockProps }/> );
  })
  it('renders a cart icon',() => {
    expect( wrapper.debug() ).toMatchSnapshot();
  })

  it('calls toggleCartHidden when clicked', () => {
    wrapper.find('CartIconContainer').simulate('click');
    expect(mockToggleCartHidden).toHaveBeenCalled();
  })
  
  it('shows the correct count in the icon',() => {
    expect(wrapper.find( 'ItemCount' ).text()).toEqual("0");
  })
})
