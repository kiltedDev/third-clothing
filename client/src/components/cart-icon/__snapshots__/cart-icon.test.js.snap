// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartIcon renders a cart icon 1`] = `
"<CartIconContainer className=\\"cart-icon\\" onClick={[Function: mockConstructor]}>
  <ShoppingIcon className=\\"shopping-icon\\" title=\\"shopping-bag\\" />
  <ItemCount className=\\"item-count\\">
    0
  </ItemCount>
</CartIconContainer>"
`;
